### ISSUE TITLE
For Open Ownership, flatten into seperate sheets could be good

### ISSUE THREAD

**Issue by odscjames** [OpenDataServices/flatten-tool#223]
**Open Ownership**

If you take the first sample from http://standard.openownership.org/en/schema-beta-2/examples/index.html you'll see it's a list of types, but different types based on the "statementType" value. Also, each type has it's own schema of what other fields are there:

```
[
  {
    "statementID": "1dc0e987-5c57-4a1c-b3ad-61353b66a9b7",
    "statementType": "entityStatement",
    "statementDate": "2017-11-18",
    "entityType": "registeredEntity",
    "name": "CHRINON LTD",
    "foundingDate": "2010-11-18",
    "identifiers": [
      {
        "scheme": "GB-COH",
        "id": "07444723"
      }
    ]
  },
  {
    "statementID": "019a93f1-e470-42e9-957b-03559861b2e2",
    "statementType": "personStatement",
    "statementDate": "2017-11-18",
    "personType": "knownPerson",
    "nationalities": [
      {
        "code": "GB"
      }
    ],
    "names": [
      {
        "type": "individual",
        "fullName": "Christopher Taggart",
        "givenName": "Christopher",
        "familyName": "Taggart"
      },
      {
        "type": "aka",
        "fullName": "Chris Taggart"
      }
    ],
    "birthDate": "1964-04",
    "addresses": [
      {
        "type": "service",
        "address": "Aston House, Cornwall Avenue, London",
        "country": "GB",
        "postCode": "N3 1LF"
      }
    ]
  },
  {
    "statementID": "fbfd0547-d0c6-4a00-b559-5c5e91c34f5c",
    "statementType": "ownershipOrControlStatement",
    "statementDate": "2017-11-18",
    "subject": {
      "describedByEntityStatement": "1dc0e987-5c57-4a1c-b3ad-61353b66a9b7"
    },
    "interestedParty": {
      "describedByPersonStatement": "019a93f1-e470-42e9-957b-03559861b2e2"
    },
    "interests": [
      {
        "type": "shareholding",
        "interestLevel": "direct",
        "beneficialOwnershipOrControl":true,
        "startDate": "2016-04-06",
        "share": {
          "exact": 100
        }
      }
    ]
  }
]
```

**Flatten**

If you use flatten on this (after applying patch #222) you'll get 6 files:

  * addresses.csv
  *  identifiers.csv
  * interests.csv
  *  main.csv
 *  names.csv
  *  nationalities.csv 

**Problem**

The problem is it's not clear which sub sheet applies to which statementType! This can make it hard to know which data is relevant in which case. Worse, if you look at main.csv

```
statementID,statementType,statementDate,entityType,name,foundingDate,personType,birthDate,subject/describedByEntityStatement,interestedParty/describedByPersonStatement
1dc0e987-5c57-4a1c-b3ad-61353b66a9b7,entityStatement,2017-11-18,registeredEntity,CHRINON LTD,2010-11-18,,,,
019a93f1-e470-42e9-957b-03559861b2e2,personStatement,2017-11-18,,,,knownPerson,1964-04,,
fbfd0547-d0c6-4a00-b559-5c5e91c34f5c,ownershipOrControlStatement,2017-11-18,,,,,,1dc0e987-5c57-4a1c-b3ad-61353b66a9b7,019a93f1-e470-42e9-957b-03559861b2e2
```

Some of the columns only apply to certain statementType's and it's not clear which. Should I fill out foundingDate or birthDate or both or none for an entry? Nothing in that spreadsheet helps me answer that! This could encourage people to put data in that isn't valid for that statement type.

**What to do about it**

It would be cool if this was split into many different sheets:

*  entityStatement.csv
*  entityStatement-identifiers.csv
 *  ownershipOrControlStatement.csv
 *  ownershipOrControlStatement-interests.csv
 *  personStatement-addresses.csv
 * personStatement.csv
 * personStatement-names.csv
 *  personStatement-nationalities.csv

Now each subsheet is clear what type it applies to. Also, the columns are only the relevant ones:

eg, entityStatement.csv

```
statementID,statementType,statementDate,entityType,name,foundingDate
1dc0e987-5c57-4a1c-b3ad-61353b66a9b7,entityStatement,2017-11-18,registeredEntity,CHRINON LTD,2010-11-18
```

eg personStatement.csv
```
statementID,statementType,statementDate,personType,birthDate
019a93f1-e470-42e9-957b-03559861b2e2,personStatement,2017-11-18,knownPerson,1964-04
```

Compare foundingDate and birthDate only being on the right sheet! Before they were both on main.

**Also, Reversible**

I think it's important that the way this is done is reversible - eg, you take a JSON, flatten it into all these sub sheets, do some work on them, and then you still need to be able to unflatten it into one JSON again. This should work fine - but test that.

**Comment by odscjames** [comment]
Note I think rollup could be used here to reduce some of the subsheets, and if that was used a lot you would end up with 3 sheets:

 *    entityStatement.csv
 *   ownershipOrControlStatement.csv
 *   personStatement.csv

This might still be easier to work with than the one "main.csv", for the 2 reasons described above.

**Comment by timgdavies** [comment]
Agreed that flattening separately by type makes good sense. 

I'm not sure if you are thinking about this as a possible 'abstract' feature, or special-case handling for BODS, but in case it is useful for generalising an approach, I could *imagine* cases in OCDS or other files where you might also want to split by some 'type' variable before flattening. For example, you might have a country where they use three kinds of contracting process ([procurementMethodDetails](http://standard.open-contracting.org/latest/en/schema/reference/#release-schema.json,/definitions/Tender,procurementMethodDetails) = ['framework', 'open', 'competitive dialogue']), and these have very different sets of fields associated with them. In this case, there could be benefit for an analyst if you you could hand a parameter to flatten tool to say 'I want you to **first** filter the top objects for their value of `tender/procurementMethodDetails`, and then flatten into a separate main sheet for each of these, **and** drop any fields that are not used in this case (see #227).

The radical rollUp approach for BODS also makes good sense - as in most cases, the one-to-many relationships are there for edge cases, rather than everyday cases.

**Comment by timgdavies** [comment]
(As an aside to note - in terms of **reversibility** - there is an extra constraint in BODS [re: ordering of statements](http://standard.openownership.org/en/schema-beta-2/schema/reference.html#serialization) that might need some thought in terms of how unflattening works - or may just need post-processing of flatten-tool output)

**Comment by odscjames** [comment]
I was thinking abstract - if you look at P.R. #224 there is a new parameter to flatten, 
--separate-by-key. Pass a field name, and for each different value it will make a different set of files. I think the rest should work out - need to think about names and edge cases tho.

**Comment by odscjames** [comment]
Discussion with @kindly - another option is a:

-  a filter option, so you can run flatten and say "only process data with X=Y". With this, you would run the command 3 times, one for each statementType
-  a prefix option, a simple flag saying "put this string infront of each file name". You can use this to avoid renaming files by hand.
-  this works for CSV's, but for Excel you want an option to add sheets to existing files. 

Noted with this you would want a 4th run; "process data with X!=Y, A or B" to catch any data with bad  statementTypes. Think about this. More work than separating in one run, but maybe more useful?

**Comment by odscjames** [comment]
Discussion with @kindly  and agreed approach:

Basic filters, 2 new command line arguments:
  *  filter field
  *  filter value

First version, field will be top level fields only but in future this could be JSON paths.

Also, add a new command line argument
  *  file prefix
that automatically prefixes all files with a string.

**Comment by jpmckinney** [comment]
Is this closed by #235 and #240?
